presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -0,0 +1,762 @@
+        long[] longs = new long[NUMBER_OF_LONGS];
+        if (unscaledValue < 0) {
+            longs[0] = -unscaledValue;
+            longs[1] = SIGN_LONG_MASK;
+        }
+        else {
+            longs[0] = unscaledValue;
+        }
+        return Slices.wrappedLongArray(longs);
+    }
+
+    public static BigInteger unscaledDecimalToBigInteger(Slice decimal)
+    {
+        byte[] bytes = decimal.getBytes(0, UNSCALED_DECIMAL_128_SLICE_LENGTH);
+        // convert to big-endian order
+        reverse(bytes);
+        bytes[0] &= ~SIGN_BYTE_MASK;
+        return new BigInteger(isNegative(decimal) ? -1 : 1, bytes);
+    }
+
+    public static long unscaledDecimalToUnscaledLong(Slice decimal)
+    {
+        long lo = getLong(decimal, 0);
+        long hi = getLong(decimal, 1);
+        boolean negative = isNegative(decimal);
+
+        if (hi != 0 || ((lo > Long.MIN_VALUE || !negative) && lo < 0)) {
+            throwOverflowException();
+        }
+
+        return negative ? -lo : lo;
+    }
+
+    public static long unscaledDecimalToUnscaledLongUnsafe(Slice decimal)
+    {
+        long lo = getLong(decimal, 0);
+        return isNegative(decimal) ? -lo : lo;
+    }
+
+    public static Slice rescale(Slice decimal, int rescaleFactor)
+    {
+        if (rescaleFactor == 0) {
+            return decimal;
+        }
+        else {
+            Slice result = unscaledDecimal();
+            rescale(decimal, rescaleFactor, result);
+            return result;
+        }
+    }
+
+    public static void rescale(Slice decimal, int rescaleFactor, Slice result)
+    {
+        if (rescaleFactor == 0) {
+            copyUnscaledDecimal(decimal, result);
+        }
+        else if (rescaleFactor > 0) {
+            if (rescaleFactor >= POWERS_OF_TEN.length) {
+                throwOverflowException();
+            }
+            multiply(decimal, POWERS_OF_TEN[rescaleFactor], result);
+        }
+        else {
+            scaleDownRoundUp(decimal, -rescaleFactor, result);
+        }
+    }
+
+    public static Slice multiply(Slice left, Slice right)
+    {
+        Slice result = unscaledDecimal();
+        multiply(left, right, result);
+        return result;
+    }
+
+    public static void multiply(Slice left, Slice right, Slice result)
+    {
+        int l0 = getInt(left, 0);
+        int l1 = getInt(left, 1);
+        int l2 = getInt(left, 2);
+        int l3 = getInt(left, 3);
+
+        int r0 = getInt(right, 0);
+        int r1 = getInt(right, 1);
+        int r2 = getInt(right, 2);
+        int r3 = getInt(right, 3);
+
+        // the combinations below definitely result in an overflow
+        if (((r3 != 0 && (l3 | l2 | l1) != 0) || (r2 != 0 && (l3 | l2) != 0) || (r1 != 0 && l3 != 0))) {
+            throwOverflowException();
+        }
+
+        int z0 = 0;
+        int z1 = 0;
+        int z2 = 0;
+        int z3 = 0;
+        int z4 = 0;
+        int z5 = 0;
+        int z6 = 0;
+        int z7 = 0;
+
+        if (l0 != 0) {
```

Comment:
Does it matter? There's no magic to it -- it's just implementing the "japanese multiplication method". It's, actually, equivalent to the following (except for accumulating the carry at each step to avoid overflow):

```java
long a0 = r0 * l0;
long a1 = r0 * l1 + r1 * l0 + (a0 >>> 32);
long a2 = r0 * l2 + r1 * l1 + r2 * l0 + (a1 >>> 32);
long a3 = r0 * l3 + r1 * l2 + r2 * l1 + r3 * l0 + (a2 >>> 32);
long a4 = r1 * l3 + r2 * l2 + r3 * l1 + (a3 >>> 32);
long a5 = r2 * l3 + r3 * l2 + (a4 >>> 32);
long a6 = r3 * l3 + (a5 >>> 32);

int z0 = (int) a0;
int z1 = (int) a1;
int z2 = (int) a2;
int z3 = (int) a3;
int z4 = (int) a4;
int z5 = (int) a5;
int z6 = (int) a6;
int z7 = (int) (a6 >>> 32);
```